src/jit/decompose.cpp

```diff
@@ -0,0 +1,1189 @@
+
+    // Decompose new node
+    DecomposeNode(treePtr, data);
+}
+
+
+//------------------------------------------------------------------------
+// DecomposeStoreInd: Decompose GT_STOREIND.
+//
+// Arguments:
+//    tree - the tree to decompose
+//
+// Return Value:
+//    None.
+//
+void Decompose::DecomposeStoreInd(GenTree** ppTree, Compiler::fgWalkData* data)
+{
+    assert(ppTree != nullptr);
+    assert(*ppTree != nullptr);
+    assert(data != nullptr);
+    assert((*ppTree)->OperGet() == GT_STOREIND);
+
+    GenTree* tree = *ppTree;
+
+    assert(tree->gtOp.gtOp2->OperGet() == GT_LONG);
+
+    GenTreeStmt* curStmt = comp->compCurStmt->AsStmt();
+    bool isEmbeddedStmt = !curStmt->gtStmtIsTopLevel();
+
+    // Example input trees (a nested embedded statement case)
+    //
+    //   <linkBegin Node>
+    //   *  stmtExpr  void  (top level) (IL   ???...  ???)
+    //   |  /--*  argPlace  ref    $280
+    //   |  +--*  argPlace  int    $4a
+    //   |  |  {  *  stmtExpr  void  (embedded) (IL   ???...  ???)
+    //   |  |  {  |     /--*  lclVar    ref    V11 tmp9         u:3 $21c
+    //   |  |  {  |     +--*  const     int    4 $44
+    //   |  |  {  |  /--*  +         byref  $2c8
+    //   |  |  {  |  |     {  *  stmtExpr  void  (embedded) (IL   ???...  ???)
+    //   |  |  {  |  |     {  |  /--*  lclFld    long   V01 arg1         u:2[+8] Fseq[i] $380
+    //   |  |  {  |  |     {  \--*  st.lclVar long  (P) V21 cse8
+    //   |  |  {  |  |     {  \--*    int    V21.hi (offs=0x00) -> V22 rat0    
+    //   |  |  {  |  |     {  \--*    int    V21.hi (offs=0x04) -> V23 rat1    
+    //   |  |  {  |  |  /--*  lclVar    int    V22 rat0          $380
+    //   |  |  {  |  |  +--*  lclVar    int    V23 rat1
+    //   |  |  {  |  +--*  gt_long   long
+    //   |  |  {  \--*  storeIndir long
+    //   |  +--*  lclVar    ref    V11 tmp9         u:3 (last use) $21c
+    //   |  +--*  lclVar    ref    V02 tmp0         u:3 $280
+    //   |  +--*  const     int    8 $4a
+    //   \--*  call help void   HELPER.CORINFO_HELP_ARRADDR_ST $205
+    //  <linkEndNode>
+
+    GenTree* linkBegin = comp->fgGetFirstNode(tree)->gtPrev;
+    GenTree* linkEnd = tree->gtNext;
+    GenTree* gtLong = tree->gtOp.gtOp2;
+
+    // Save address to a temp. It is used in storeIndLow and storeIndHigh trees.
+    GenTreeStmt* addrStmt = CreateEmbeddedFormTemp(&tree->gtOp.gtOp1);
+    JITDUMP("[DecomposeStoreInd]: Saving address tree to a temp var:\n");
+    DISPTREE(addrStmt);
+
+    if (!gtLong->gtOp.gtOp1->OperIsLeaf())
+    {
+        GenTreeStmt* dataLowStmt = CreateEmbeddedFormTemp(&gtLong->gtOp.gtOp1);
+        JITDUMP("[DecomposeStoreInd]: Saving low data tree to a temp var:\n");
+        DISPTREE(dataLowStmt);
+    }
+
+    if (!gtLong->gtOp.gtOp2->OperIsLeaf())
+    {
+        // REVIEW: should we also use CreateEmbeddedFormTemp() here?
+        GenTreeStmt* dataHighStmt = comp->fgInsertEmbeddedFormTemp(&gtLong->gtOp.gtOp2);
```

Comment:
> should we also use CreateEmbeddedFormTemp() here?

IMO yes. As explained in the `GT_NEG` PR that logic isn't necessary here but I don't think it helps to avoid using `CreateEmbeddedFormTemp`, quite the contrary.